lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx

```diff
@@ -82,6 +94,899 @@ GeneralizedExtremeValue GeneralizedExtremeValueFactory::buildAsGeneralizedExtrem
+};
+
+/* R largest order statistics */
+DistributionFactoryResult GeneralizedExtremeValueFactory::buildRMaximaEstimator(const Sample & sample,
+                                                                                const UnsignedInteger rx)
+{
+  const UnsignedInteger R = sample.getDimension();
+  // r=0 means r=R
+  const UnsignedInteger r = rx ? rx : R;
+  const UnsignedInteger size = sample.getSize();
+  if (r > R)
+    throw InvalidArgumentException(HERE) << "r(" << r << ") should be < R (" << R << ")";
+  if (size < 2)
+    throw InvalidArgumentException(HERE) << "Error: can build a GeneralizedExtremeValue distribution only from a sample of size>=2, here size=" << sample.getSize();
+  const UnsignedInteger m = sample.getSize();
+
+  for (UnsignedInteger i = 0; i < m; ++ i)
+    for (UnsignedInteger j = 0; j < r - 1; ++ j)
+      if (sample(i, j) < sample(i, j + 1))
+        throw InvalidArgumentException(HERE) << "The maxima of bloc #" << (i + 1) << "/" << m << " are not sorted in decreasing order";
+
+  const Function objective(new GeneralizedExtremeValueRMaximaLikelihoodEvaluation(sample, r));
+  OptimizationProblem problem(objective);
+  problem.setMinimization(false);
+
+  // sigma > 0
+  const Point lowerBound({-SpecFunc::MaxScalar, 0.0, -SpecFunc::MaxScalar});
+  const Point upperBound(3, SpecFunc::MaxScalar);
+  const Interval::BoolCollection finiteLowerBound({false, true, false});
+  const Interval::BoolCollection finiteUpperBound(3, false);
+  problem.setBounds(Interval(lowerBound, upperBound, finiteLowerBound, finiteUpperBound));
+
+  // pwm + 1d search for the starting point, see gevrFit function from R eva package
+  const Sample rMax(sample.sort(0));
+  Sample rMax1(rMax);
+  Sample rMax2(rMax);
+  for (UnsignedInteger i = 0; i < m; ++ i)
+  {
+    rMax1(i, 0) *= 1.0 * i / (m - 1.0);
+    rMax2(i, 0) *= 1.0 * i * (i - 1.0) / (m - 1.0) / (m - 2.0);
+  }
+  const Scalar mom0 = rMax.computeMean()[0];
+  const Scalar mom1 = rMax1.computeMean()[0];
+  const Scalar mom2 = rMax2.computeMean()[0];
+  const SymbolicFunction solveShape(Description({"sh", "mom0", "mom1", "mom2"}),
+                                    Description(1, "(3^sh - 1) / (2^sh - 1) - (3 * mom2 - mom0) / (2 * mom1 - mom0)"));
+  const ParametricFunction f(solveShape, Indices({1, 2, 3}), Point({mom0, mom1, mom2}));
+  const Scalar a = ResourceMap::GetAsScalar("GeneralizedExtremeValueFactory-XiSearchLowerBound");
+  const Scalar b = ResourceMap::GetAsScalar("GeneralizedExtremeValueFactory-XiSearchUpperBound");
+  const Scalar fA = f(Point(1, a))[0];
+  const Scalar fB = f(Point(1, b))[0];
+  Brent solver1d;
+  const Scalar xi0 = solver1d.solve(f, 0.0, a, b, fA, fB);
+  const Scalar sigma0 = ((2.0 * mom1 - mom0) * xi0) / (SpecFunc::Gamma(1.0 - xi0) * (std::pow(2.0, xi0) - 1.0));
+  const Scalar mu0 = mom0 + sigma0 * (1.0 - SpecFunc::Gamma(1.0 - xi0)) / xi0;
+  const Point x0({mu0, sigma0, xi0});
+
+  // solve optimization problem
+  TNC solver(problem);
```

Comment:
TNC is not the most reliable algorithm...

Reply:
yes, but it turned out to be best here